pkg/template/processor.go

```diff
@@ -0,0 +1,156 @@
+package template
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"sort"
+	"time"
+
+	projectv1 "github.com/openshift/api/project/v1"
+	templatev1 "github.com/openshift/api/template/v1"
+	"github.com/openshift/library-go/pkg/template/generator"
+	"github.com/openshift/library-go/pkg/template/templateprocessing"
+	errs "github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type Processor struct {
+	cl           client.Client
+	scheme       *runtime.Scheme
+	codecFactory serializer.CodecFactory
+}
+
+func NewProcessor(cl client.Client, scheme *runtime.Scheme) *Processor {
+	return &Processor{cl: cl, scheme: scheme, codecFactory: serializer.NewCodecFactory(scheme)}
+}
+
+func (p Processor) ProcessAndApply(tmplContent []byte, values map[string]string) error {
+	objs, err := p.Process(tmplContent, values)
+	if err != nil {
+		return errs.Wrap(err, "failed while processing template")
+	}
+
+	return p.apply(objs)
+}
+
+func (p Processor) apply(objs []runtime.RawExtension) error {
+	// sorting before applying to maintain correct order
+	sort.Sort(ByKind(objs))
+
+	for _, rawObj := range objs {
+		obj := rawObj.Object
+		if obj == nil {
+			continue
+		}
+		gvk := obj.GetObjectKind().GroupVersionKind()
+		if err := createOrUpdateObj(p.cl, obj); err != nil {
+			return errs.Wrapf(err, "unable to create resource of kind: %s, version: %s", gvk.Kind, gvk.Version)
+		}
+		if gvk.Group == "project.openshift.io" && gvk.Version == "v1" && gvk.Kind == "ProjectRequest" {
+			var prq projectv1.ProjectRequest
+			err := p.scheme.Convert(obj, &prq, nil)
+			if err != nil {
+				return errs.Wrapf(err, "unable to convert object of kind: %s, version: %s to a ProjectRequests", gvk.Kind, gvk.Version)
+			}
+			// wait until the Project exists and is ready
+			_, err = p.waitUntilProjectExists(prq.GetName(), 10*time.Second)
```

Comment:
I'm a bit concerned about the fact that we wait 10 seconds for the project to be created and active. What if it takes longer for whatever reason? It would require proper alignment with the NSTemplateSet controller logic which is to use this template processor.
But TBH I don't know how we could do better without introducing too much of complexity.

Reply:
in e4cc320 I've added a param to the `Apply` func to configure the project creation timeout, so it's up to the controller to decide how long to wait. I suggest we externalize this setting in the deployment environment (https://github.com/codeready-toolchain/member-operator/blob/master/deploy/operator.yaml#L23), but let's do that in a separate PR, ok?

Reply:
Sure. Having a better way of controlling timeouts and delays is useful but it doesn't eliminate the potential problem with too long namespace/project creation completely. We would need to drop our own custom waits and relay on controller reconcile events for that. But as I said it would probably bring a lot of complexity in our template process flow.
So, I'm fine with using that wait for now but wanted to share my overall concern with such approach :)